### PR TITLE
Fix broken bazel rules

### PR DESCRIPTION
**Pull request by fejta**
/assign @Katharine @cjwagner 

```console
b0ebb806f871834291964fdb37490052/external/io_bazel_rules_appengine/appengine/py_appengine.bzl", line 149, in attr.label_list
		FileType([".yaml", ".py"])
FileType function is not available. You may use a list of strings instead. You can temporarily reenable the function by passing the flag --incompatible_disallow_filetype=false
ERROR: error loading package '': Extension file 'appengine/py_appengine.bzl' has errors
```

```console
prow/BUILD.bazel", line 14, in prow_push
		tags("artifact-uploader", "branchprotec...", <20 more arguments>) + image_tags(**{prefix("needs-rebase"): "//pr..."})

The `+` operator for dicts is deprecated and no longer supported. Please use the `update` method instead. You can temporarily enable the `+` operator by passing the flag --incompatible_disallow_dict_plus=false
```

### CODE REVIEW

**Comment by Katharine** [comment]
/lgtm

but looks like we need bazel 0.23+ for these changes to work.

**Comment by stevekuznetsov** [comment]
Will this also require all devs to update bazel

**Comment by Katharine** [comment]
Yup.

**Comment by stevekuznetsov** [comment]
Let's blast that out to the mailing list and Slack etc :)

File: prow/BUILD.bazel [hunk 1]
```diff
@@ -35,19 +40,8 @@ prow_push(
...
-    }) + image_tags(**{
-        prefix("refresh"): "//prow/external-plugins/refresh:image",
-    }) + image_tags(**{
-        prefix("ghproxy"): "//ghproxy:image",
-    }) + image_tags(**{
-        prefix("label_sync"): "//label_sync:image",
-    }) + image_tags(**{
-        prefix("commenter"): "//robots/commenter:image",
-    }),
+        **_other_images,
```
Comment (ixdy) [inline prow/BUILD.bazel]: you could also just pass this as a named parameter instead?

Comment (fejta) [reply prow/BUILD.bazel]: That screws up *args

**Comment by fejta** [comment]
Once the above issues are merged I can use a better version of kazel

**Comment by fejta** [comment]
Passes tests now!

File: WORKSPACE [hunk 2]
```diff
@@ -470,9 +470,9 @@ py_library(
 # TODO(fejta): get this to work
 git_repository(
     name = "io_bazel_rules_appengine",
-    commit = "14d860985c2a764fdb6a6072d5450d8360c4ce5b",
+    commit = "fdbce051adecbb369b15260046f4f23684369efc",
     remote = "https://github.com/bazelbuild/rules_appengine.git",
-    #tag = "0.0.5", # Latest at https://github.com/bazelbuild/rules_appengine/releases.
+    #tag = "0.0.8+but-this-isn't-new-enough", # Latest at https://github.com/bazelbuild/rules_appengine/releases.
```
Comment (cjwagner) [inline WORKSPACE]: What does this mean?

File: prow/def.bzl [hunk 3]
```diff
@@ -72,38 +71,43 @@ BUILD_CLUSTER = "{STABLE_BUILD_CLUSTER}"  # For untrusted test code
...
+#     "gcr.io/k8s-prow/ghproxy:latest-fejta": "//ghproxy:image",
 #   }
-def tags(*cmds):
-  # Create :YYYYmmdd-commitish :latest :latest-USER tags
-  return _docker_tags(**{prefix(cmd): target(cmd) for cmd in cmds})
+def tags(*cmds, **targets):
+    # Create :YYYYmmdd-commitish :latest :latest-USER tags
+    cmd_targets = {prefix(cmd): target(cmd) for cmd in cmds}
+    cmd_targets.update({prefix(p): t for (p, t) in targets.items()})
+    return _image_tags(**{prefix(cmd): target(cmd) for cmd in cmds})
```
Comment (cjwagner) [inline prow/def.bzl]: `cmd_targets` is never used?

File: hack/update-bazel.sh [hunk 4]
```diff
@@ -19,6 +19,10 @@ set -o pipefail
 
 cd "$(git rev-parse --show-toplevel)"
 mkdir -p ./vendor
-touch ./vendor/BUILD.bazel
+if [[ ! -e ./vendor/BUILD.bazel ]]; then
+  echo "Bootstrapping vendor..." >&2
+  touch ./vendor/BUILD.bazel
+  bazel run //:gazelle-bootstrap
```
Comment (cjwagner) [inline hack/update-bazel.sh]: Why do we do this only when `./vendor/BUILD.bazel` does not exist? What if `./vendor` is updated in a way that necessitates a change to the BUILD files?

File: WORKSPACE (hunk 2 above)
Comment (fejta) [reply WORKSPACE]: The above commit is after 0.0.8, as 0.0.8 remains broken, but head is fixed. However 0.0.8 is the latest release

File: hack/update-bazel.sh (hunk 4 above)
Comment (fejta) [reply hack/update-bazel.sh]: If someone updates something under in vendor, then the `bazel run //:gazelle` command will fix everything in vendor up.

If is extremely odd if `./vendor/BUILD.bazel` doesn't exist. This most likely means that we just recreated the vendor dir during an `update-deps.sh` which will `rm -rf vendor && go mod vendor`. At this time, there are no BUILD.bazel files, so we cannot use the vendored version of gazelle. Instead we do an initial pass with the (potentially wrong) version included with the bazel-gazelle WORKSPACE repo rule. This creates the vendored gazelle targets.

Right after this we run `bazel run //:gazelle` which will use the vendored version of gazelle to ensure we're generating things correctly.

Normally we don't delete vendor before running `hack/update-bazel.sh` so vendored gazelle rule already exists so we don't need to bootstrap it.

File: prow/def.bzl (hunk 3 above)
Comment (fejta) [reply prow/def.bzl]: Done

**Comment by cjwagner** [comment]
/hold
For https://github.com/kubernetes/test-infra/pull/11987#issuecomment-477765782

**Comment by Katharine** [comment]
/lgtm

(I don't think it makes much sense to hold this until _after_ telling people to update, because if they update before this merges they get incomprehensible errors until this merges - while if this merges before they update they get an error telling them that their bazel version is too old.)

**Comment by fejta** [comment]
/hold cancel

Sent out the mail

**Pull request merged**